addons/voice/org.openhab.voice.gtts/src/main/java/org/openhab/voice/gtts/internal/GoogleCloudAPI.java

```diff
@@ -0,0 +1,350 @@
+     */
+    private InputStream synthesizeSpeechByGoogle(String text, GoogleTTSVoice voice, String audioFormat) {
+        // Set the text input to be synthesized
+        SynthesisInput.Builder builder = SynthesisInput.newBuilder();
+        if (text.startsWith("<speak>")) {
+            builder.setSsml(text);
+        } else {
+            builder.setText(text);
+        }
+        SynthesisInput input = builder.build();
+
+        // Build the voice request
+        VoiceSelectionParams voiceParam = VoiceSelectionParams.newBuilder()
+                .setLanguageCode(voice.getLocale().getLanguage())
+                .setName(voice.getLabel())
+                .setSsmlGender(SsmlVoiceGender.forNumber(voice.getSsmlGender()))
+                .build();
+
+        // Select the type of audio file you want returned
+        AudioConfig audioConfig = AudioConfig.newBuilder()
+                .setAudioEncoding(AudioEncoding.valueOf(audioFormat)) // MP3 or LINEAR16 audio.
+                .setPitch(config.getPitch())
+                .setSpeakingRate(config.getSpeakingRate())
+                .setVolumeGainDb(config.getVolumeGainDb())
+                .build();
+
+        // Perform the text-to-speech request
+        SynthesizeSpeechResponse response = googleClient.synthesizeSpeech(input, voiceParam,
+                audioConfig);
+
+        // Get the audio contents from the response
+        ByteString audioContents = response.getAudioContent();
+
+        return new ByteArrayInputStream(audioContents.toByteArray());
+    }
+
+    /**
+     * Gets a unique filename for a give text, by creating a MD5 hash of it. It
+     * will be preceded by the locale.
+     * <p>
+     * Sample: "en-US_00a2653ac5f77063bc4ea2fee87318d3"
+     */
+    private String getUniqueFilenameForText(String text, Locale locale) {
+        try {
+            byte[] bytesOfMessage = text.getBytes("UTF-8");
+            MessageDigest md = MessageDigest.getInstance("MD5");
+            byte[] md5Hash = md.digest(bytesOfMessage);
+            BigInteger bigInt = new BigInteger(1, md5Hash);
+            StringBuilder hashText = new StringBuilder(bigInt.toString(16));
+            // Now we need to zero pad it if you actually want the full 32
+            // chars.
+            while (hashText.length() < 32) {
+                hashText.insert(0, "0");
+            }
+            return locale.getLanguage() + "_" + hashText;
+        } catch (UnsupportedEncodingException | NoSuchAlgorithmException ex) {
+            // should not happen
+            logger.error("Could not create MD5 hash for '{}'", text, ex);
+            return null;
+        }
+    }
+
+    private void copyStream(InputStream inputStream, OutputStream outputStream) throws IOException {
+        byte[] bytes = new byte[4096];
+        int read = inputStream.read(bytes, 0, 4096);
+        while (read > 0) {
+            outputStream.write(bytes, 0, read);
+            read = inputStream.read(bytes, 0, 4096);
```

Comment:
Extracting `4096` to a constant or using `bytes.length` will prevent future issues when one day the values are changed but one is forgotten or ends up with a typo.

Reply:
Done